scripts/language_model/large_word_language_model.py

```diff
@@ -0,0 +1,327 @@
+
+            if nbatch % args.log_interval == 0:
+                cur_L = total_L / args.log_interval / len(context)
+                ppl = math.exp(cur_L) if cur_L < 100 else 99999999
+                print('[Epoch %d Batch %d] loss %.2f, ppl %.2f, '
+                      'throughput %.2f samples/s'
+                      %(epoch, nbatch, cur_L, ppl,
+                        train_batch_size*args.log_interval/(time.time()-start_log_interval_time)))
+                total_L = 0.0
+                start_log_interval_time = time.time()
+        end_epoch_time = time.time()
+        print('Epoch %d took %.2f seconds.'%(epoch, end_epoch_time - start_epoch_time))
+        mx.nd.waitall()
+        checkpoint_name = '%s.%s'%(args.save, format(epoch, '02d'))
+        model.save_parameters(checkpoint_name)
+        trainer.save_states('%s.state'%args.save)
+
+def detach(hidden):
+    if isinstance(hidden, (tuple, list)):
+        hidden = [detach(h) for h in hidden]
+    else:
+        hidden = hidden.detach()
+    return hidden
+
+def test(data_stream, batch_size, ctx=None):
+    """Evaluate the model on the dataset.
+
+    Parameters
+    ----------
+    data_stream : DataStream
+        The dataset is testd on.
+    batch_size : int
+        The size of the mini-batch.
+    ctx : mx.cpu() or mx.gpu()
+        The context of the computation.
+
+    Returns
+    -------
+    loss: float
+        The loss on the dataset
+    """
+    total_L = 0.0
+    ntotal = 0
+    nbatch = 0
+    hidden = eval_model.begin_state(batch_size=batch_size, func=mx.nd.zeros, ctx=ctx)
+    for data, target, mask in data_stream:
+        data = data.as_in_context(ctx)
+        target = target.as_in_context(ctx)
+        mask = mask.as_in_context(ctx)
+        output, hidden = eval_model(data, hidden)
+        hidden = detach(hidden)
+        L = loss(output, target.reshape(-1,)) * mask.reshape((-1,))
+        total_L += L.sum()
+        ntotal += mask.sum()
+        nbatch += 1
+        avg = (total_L / ntotal).asscalar()
+        if nbatch % args.log_interval == 0:
+            print('Evaluation batch %d: test loss %.2f, test ppl %.2f'
+                  %(nbatch, avg, math.exp(avg)))
+        if max_nbatch_eval and nbatch > max_nbatch_eval:
+            print('Quit evaluation early at batch %d'%nbatch)
+            break
+    return avg
+
+def evaluate():
+    """ Evaluate loop for the trained model """
+    print(eval_model)
+    eval_model.initialize(mx.init.Xavier(), ctx=context[0])
+    eval_model.hybridize(static_alloc=True, static_shape=True)
+    epoch = args.from_epoch if args.from_epoch else 0
+    while epoch < args.epochs:
+        checkpoint_name = '%s.%s'%(args.save, format(epoch, '02d'))
+        if not os.path.exists(checkpoint_name):
+            print('Wait for a new checkpoint...')
+            # check again after 600 seconds
+            time.sleep(600)
+            continue
+        eval_model.load_parameters(checkpoint_name)
```

Comment:
Instead of waiting for another thread to write the parameters to a file, would it make sense to extend the `Block.load_parameters` to take another Block as an input? Then you may save the roundtrip via disk and potential issues with waiting for the writer thread to finish.

Reply:
As discussed offline:
For training, usually SampledLogits block is recommended if there's enough memory on GPU. If not, we need to use SparseSampledLogits to reduce memory consumption. 
If there's already not enough memory for training, we will not run evaluation in the same script, since evaluation requires full parameters for the last layer and computes the full softmax. The user is expected to take a checkpoint and run another script to run evaluation on cpu in parallel.